File: src/Modules/SmartPrompt/Contracts/BehaviorModels.cs
namespace DigitalBrain.SmartPrompt;

public enum BehaviorDiagnosticSeverity
{
    Info,
    Warning,
    Error,
}

public enum BehaviorStepRole
{
    Setup,
    Trigger,
    Filter,
    Action,
    Fake,
    Invoke,
    Assert,
}

[GenerateSerializer]
[Alias("db.behavior.diagnostic.v1")]
public sealed record BehaviorDiagnostic(
    [property: Id(0)] string Code,
    [property: Id(1)] BehaviorDiagnosticSeverity Severity,
    [property: Id(2)] string Message,
    [property: Id(3)] int Line,
    [property: Id(4)] int Column);

[GenerateSerializer]
[Alias("db.behavior.step-call.v1")]
public sealed record BehaviorStepCall(
    [property: Id(0)] string Keyword,
    [property: Id(1)] string Text,
    [property: Id(2)] string Binding,
    [property: Id(3)] BehaviorStepRole Role,
    [property: Id(4)] IReadOnlyList<string> Arguments,
    [property: Id(5)] int Line);

[GenerateSerializer]
[Alias("db.behavior.scenario-plan.v1")]
public sealed record BehaviorScenarioPlan(
    [property: Id(0)] string Name,
    [property: Id(1)] string TriggerKey,
    [property: Id(2)] IReadOnlyList<BehaviorStepCall> Steps);

[GenerateSerializer]
[Alias("db.behavior.test-plan.v1")]
public sealed record BehaviorTestPlan(
    [property: Id(0)] string Name,
    [property: Id(1)] IReadOnlyList<BehaviorStepCall> Steps);

[GenerateSerializer]
[Alias("db.behavior.plan.v1")]
public sealed record BehaviorPlan(
    [property: Id(0)] string Feature,
    [property: Id(1)] string SourceHash,
    [property: Id(2)] IReadOnlyList<BehaviorScenarioPlan> Behaviors,
    [property: Id(3)] IReadOnlyList<BehaviorTestPlan> Tests);

[GenerateSerializer]
[Alias("db.behavior.compilation.v1")]
public sealed record BehaviorCompilation(
    [property: Id(0)] BehaviorPlan? Plan,
    [property: Id(1)] IReadOnlyList<BehaviorDiagnostic> Diagnostics)
{
    public bool Success => Plan is not null
        && Diagnostics.All(static diagnostic => diagnostic.Severity != BehaviorDiagnosticSeverity.Error);
}

[GenerateSerializer]
[Alias("db.behavior.step-suggestion.v1")]
public sealed record BehaviorStepSuggestion(
    [property: Id(0)] string Keyword,
    [property: Id(1)] string Template,
    [property: Id(2)] string Description);

public interface IBehaviorCompiler
{
    BehaviorCompilation Compile(string source);

    IReadOnlyList<BehaviorStepSuggestion> Suggestions { get; }
}

[GenerateSerializer]
[Alias("db.behavior.generation.v1")]
public sealed record BehaviorGeneration(
    [property: Id(0)] string Source,
    [property: Id(1)] BehaviorCompilation Compilation,
    [property: Id(2)] string Model);

public interface IBehaviorFeatureGenerator
{
    Task<BehaviorGeneration> Generate(string request, CancellationToken cancellationToken = default);
}

[GenerateSerializer]
[Alias("db.behavior.summary.v1")]
public sealed record BehaviorSummary(
    [property: Id(0)] string Name,
    [property: Id(1)] string Title,
    [property: Id(2)] string Source,
    [property: Id(3)] bool Active,
    [property: Id(4)] BehaviorTestReport? LastTest,
    [property: Id(5)] IReadOnlyList<BehaviorDiagnostic> Diagnostics);

[GenerateSerializer]
[Alias("db.behavior.event.v1")]
public sealed record BehaviorEvent(
    [property: Id(0)] string EventId,
    [property: Id(1)] string Kind,
    [property: Id(2)] string Source,
    [property: Id(3)] string Text,
    [property: Id(4)] double Value,
    [property: Id(5)] string SourceUri,
    [property: Id(6)] DateTimeOffset OccurredAt)
{
    public string TriggerKey => Kind.Equals("x.post", StringComparison.OrdinalIgnoreCase)
        ? $"x.post/account:{Source.Trim().ToLowerInvariant()}"
        : $"{Kind.Trim().ToLowerInvariant()}/source:{Source.Trim().ToLowerInvariant()}";
}

[GenerateSerializer]
[Alias("db.behavior.subscription.v1")]
public sealed record BehaviorSubscription(
    [property: Id(0)] string Owner,
    [property: Id(1)] string BehaviorName,
    [property: Id(2)] string ScenarioName,
    [property: Id(3)] string RevisionHash);

[GenerateSerializer]
[Alias("db.behavior.test-report.v1")]
public sealed record BehaviorTestReport(
    [property: Id(0)] bool AllGreen,
    [property: Id(1)] IReadOnlyList<string> Failures,
    [property: Id(2)] int Scenarios);

[GenerateSerializer]
[Alias("db.behavior.definition-state.v1")]
public sealed record BehaviorDefinitionState(
    [property: Id(0)] string Source,
    [property: Id(1)] BehaviorCompilation Compilation,
    [property: Id(2)] bool Active,
    [property: Id(3)] BehaviorTestReport? LastTest);

[GenerateSerializer]
[Alias("db.behavior.catalog-state.v1")]
public sealed record BehaviorCatalogState(
    [property: Id(0)] IReadOnlyList<string> Names);

[GenerateSerializer]
[Alias("db.behavior.directory-stats.v1")]
public sealed record BehaviorDirectoryStats(
    [property: Id(0)] int SubscriptionCount,
    [property: Id(1)] int ActivePartitions,
    [property: Id(2)] int ConfiguredPartitions);

public static class BehaviorRouting
{
    public const int PartitionCount = 64;
}

public static class BehaviorIngressNames
{
    public const string Shared = "shared";
}

File: src/Modules/UI/Flutter/core/lib/src/behavior_models.dart
final class BehaviorTestReport {
  const BehaviorTestReport({
    required this.allGreen,
    required this.scenarios,
    required this.failures,
  });

  factory BehaviorTestReport.fromJson(Map<String, Object?> json) =>
      BehaviorTestReport(
        allGreen: json['allGreen'] as bool? ?? false,
        scenarios: (json['scenarios'] as num?)?.toInt() ?? 0,
        failures: (json['failures'] as List<Object?>? ?? const [])
            .cast<String>(),
      );

  final bool allGreen;
  final int scenarios;
  final List<String> failures;
}

final class BehaviorDiagnostic {
  const BehaviorDiagnostic({
    required this.code,
    required this.message,
    required this.line,
    required this.severity,
  });

  factory BehaviorDiagnostic.fromJson(Map<String, Object?> json) =>
      BehaviorDiagnostic(
        code: json['code'] as String? ?? '',
        message: json['message'] as String? ?? '',
        line: (json['line'] as num?)?.toInt() ?? 0,
        severity: json['severity']?.toString() ?? '0',
      );

  final String code;
  final String message;
  final int line;
  final String severity;
}

final class BehaviorSummary {
  const BehaviorSummary({
    required this.name,
    required this.title,
    required this.source,
    required this.active,
    required this.diagnostics,
    this.lastTest,
  });

  factory BehaviorSummary.fromJson(Map<String, Object?> json) =>
      BehaviorSummary(
        name: json['name'] as String,
        title: json['title'] as String,
        source: json['source'] as String,
        active: json['active'] as bool? ?? false,
        lastTest: json['lastTest'] is Map<String, Object?>
            ? BehaviorTestReport.fromJson(
                json['lastTest']! as Map<String, Object?>,
              )
            : null,
        diagnostics: (json['diagnostics'] as List<Object?>? ?? const [])
            .cast<Map<String, Object?>>()
            .map(BehaviorDiagnostic.fromJson)
            .toList(),
      );

  final String name;
  final String title;
  final String source;
  final bool active;
  final BehaviorTestReport? lastTest;
  final List<BehaviorDiagnostic> diagnostics;

  BehaviorSummary copyWith({
    String? source,
    bool? active,
    BehaviorTestReport? lastTest,
    List<BehaviorDiagnostic>? diagnostics,
  }) => BehaviorSummary(
    name: name,
    title: title,
    source: source ?? this.source,
    active: active ?? this.active,
    lastTest: lastTest ?? this.lastTest,
    diagnostics: diagnostics ?? this.diagnostics,
  );
}

final class BehaviorStepSuggestion {
  const BehaviorStepSuggestion({
    required this.keyword,
    required this.template,
    required this.description,
  });

  factory BehaviorStepSuggestion.fromJson(Map<String, Object?> json) =>
      BehaviorStepSuggestion(
        keyword: json['keyword'] as String,
        template: json['template'] as String,
        description: json['description'] as String,
      );

  final String keyword;
  final String template;
  final String description;
}

final class BehaviorGeneration {
  const BehaviorGeneration({
    required this.source,
    required this.model,
    required this.success,
    required this.diagnostics,
  });

  factory BehaviorGeneration.fromJson(Map<String, Object?> json) {
    final compilation = json['compilation']! as Map<String, Object?>;
    return BehaviorGeneration(
      source: json['source'] as String,
      model: json['model'] as String,
      success: compilation['success'] as bool? ?? false,
      diagnostics: (compilation['diagnostics'] as List<Object?>? ?? const [])
          .cast<Map<String, Object?>>()
          .map(BehaviorDiagnostic.fromJson)
          .toList(),
    );
  }

  final String source;
  final String model;
  final bool success;
  final List<BehaviorDiagnostic> diagnostics;
}
